src/pubsub/src/publisher/publisher.rs

```diff
@@ -153,28 +156,100 @@ impl Worker {
     }
 
     async fn run(mut self) {
-        while let Some(msg) = self.rx.recv().await {
-            let client = self.client.clone();
-            let topic = self.topic_name.clone();
-            // In the future, we may also want to keep track of JoinHandles in order to
-            // flush the results.
-            let _handle = tokio::spawn(async move {
-                // For now, we just send the message immediately.
-                // We will want to batch these requests.
-                let request = client
-                    .publish()
-                    .set_topic(topic)
-                    .set_messages(vec![msg.msg]);
-
-                // Handle the response by extracting the message ID on success.
-                let result = request
-                    .send()
-                    .await
-                    .map(|mut response| response.message_ids.pop().unwrap_or_default());
-
-                // The user may have dropped the handle, so it is ok if this fails.
-                let _ = msg.tx.send(result);
-            });
+        let mut batch = Batch::new();
+        // Set max delay at 1 year to avoid overflow. This should probably be in the setting itself.
+        // Would be reasonable to make it much less.
+        let delay = self.batching_options.delay_threshold.min(
+            Duration::from_secs(60 * 60 * 24 * 365), // 1 year.
+        );
+        let message_limit = self.batching_options.message_count_threshold;
+
+        let timer = tokio::time::sleep(delay);
+        // Pin the timer to the stack.
+        tokio::pin!(timer);
+        loop {
+            tokio::select! {
+                // Handle timer events.
+                // This branch will only be checked when there is a non-empty batch,
+                // so this will not fire continuously.
+                _ = &mut timer, if !batch.batch.is_empty() => {
+                    let batch_to_send = std::mem::take(&mut batch);
+                    tokio::spawn(batch_to_send.send(self.client.clone(), self.topic_name.clone()));
+                }
+                // Handle receiving a message from the channel.
+                msg = self.rx.recv() => {
+                    match msg {
+                        Some(msg) => {
+                            // Reset the timer if this is the first message to be added to the batch.
+                            if batch.batch.is_empty() {
```

Comment:
I find the mixing of `batch.push()` and `batch.batch.is_empty()` a bit confusing. let's try to make this consistent?